src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+// by the Apache License, Version 2.0.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+
+use mz_repr::adt::datetime::DateTimeUnits;
+use mz_repr::{ColumnType, Datum, RelationType, Row, RowArena, ScalarType};
+
+use crate::{
+    BinaryFunc, EvalError, MapFilterProject, MfpPlan, MirScalarExpr, UnaryFunc,
+    UnmaterializableFunc, VariadicFunc,
+};
+
+/// Tracks whether a function is monotonic, either increasing or decreasing.
+/// This property is useful for static analysis because a monotonic function maps ranges to ranges:
+/// ie. if `a` is between `b` and `c` inclusive, then `f(a)` is between `f(b)` and `f(c)` inclusive.
+/// (However, if either `f(b)` and `f(c)` error or return null, we don't assume anything about `f(a)`.
+///
+/// This should likely be moved to a function annotation in the future.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum Monotonic {
+    /// Monotonic in this argument.
+    Yes,
+    /// We haven't classified this one yet. (Conservatively treated as "no", but a distinct
+    /// enum so it's easy to tell whether someone's thought about a particular function yet.)
+    /// This should likely be removed when this becomes a function annotation.
+    Maybe,
+    /// We don't know any useful properties of this function.
+    No,
+}
+
+fn unary_monotonic(func: &UnaryFunc) -> Monotonic {
+    use crate::func::impls as funcs;
+    use Monotonic::*;
+    use UnaryFunc::*;
+    match func {
+        // Casts are generally monotonic.
+        CastUint64ToNumeric(_)
+        | CastNumericToUint64(_)
+        | CastNumericToMzTimestamp(_)
+        | CastUint64ToMzTimestamp(_)
+        | CastTimestampToMzTimestamp(_)
+        | CastTimestampTzToMzTimestamp(_) => Yes,
+        // Including JSON casts.
+        CastJsonbToNumeric(_)
+        | CastJsonbToBool(_)
+        | CastJsonbToString(_)
+        | CastJsonbToInt64(_)
+        | CastJsonbToFloat64(_) => Yes,
+
+        // Extracting the "most significant bits" of the a timestamp is monotonic.
+        ExtractTimestamp(funcs::ExtractTimestamp(units))
+        | ExtractTimestampTz(funcs::ExtractTimestampTz(units)) => match units {
+            DateTimeUnits::Epoch
+            | DateTimeUnits::Millennium
+            | DateTimeUnits::Century
+            | DateTimeUnits::Decade
+            | DateTimeUnits::Year => Yes,
+            _ => No,
+        },
+        DateTruncTimestamp(_) | DateTruncTimestampTz(_) => Yes,
+        // Negation is monotonically decreasing, but that's fine.
+        NegInt64(_) | NegNumeric(_) | Not(_) => Yes,
+        // Trivially monotonic, since all non-null values map to `false`.
+        IsNull(_) => Yes,
+        AbsInt64(_) => No,
+        _ => Maybe,
+    }
+}
+
+fn binary_monotonic(func: &BinaryFunc) -> (Monotonic, Monotonic) {
+    use BinaryFunc::*;
+    use Monotonic::*;
+
+    match func {
+        AddInt64 | MulInt64 | AddNumeric | MulNumeric | SubInt64 | SubNumeric => (Yes, Yes),
```

Comment:
I don't quite understand the result type of this function: do we care if the function is monotonically increasing in its first argument but decreasing in its second?

The math notion of monotonic I know would say that `-` is not monotonic, because if `x1 < x2` and `y1 < y2`, it's not necessarily the case that `x1 - y1 < x2 - y2`. Say, `10 - 7 = 3` and `6 - 1 = 5`. But considered pointwise, `-` is monotonic in its first argument and anti-monotone in its second.